spec/models/comment_spec.rb

```diff
@@ -0,0 +1,34 @@
+require "rails_helper"
+
+RSpec.describe Comment do
+  let(:restaurant_1) {FactoryBot.create :restaurant}
+  let(:user_1) {FactoryBot.create :user}
+  subject {FactoryBot.build :comment, restaurant_id: restaurant_1.id, user_id: user_1.id}
+
+  it {is_expected.to be_valid}
+
+  context "columns" do
+    it {is_expected.to have_db_column(:content).of_type(:string)}
+    it {is_expected.to have_db_column(:restaurant_id).of_type(:integer)}
+    it {is_expected.to have_db_column(:user_id).of_type(:integer)}
+  end
+
+  context "content" do
+    it {is_expected.to validate_presence_of(:content)}
+    it {is_expected.to validate_length_of(:content)}
+  end
+
+  context "associations" do
+    it {is_expected.to belong_to(:user)}
+    it {is_expected.to belong_to(:restaurant)}
+    it {is_expected.to have_many(:comment_likes).with_foreign_key(:comment_id)}
+  end
+
+  context "scope" do
+    let!(:comment_1) {FactoryBot.create :comment, restaurant_id: restaurant_1.id, user_id: user_1.id, created_at: "2019-06-07 01:41:41"}
```

Comment:
phân biệt let vs let!

Reply:
Dạ anh
+ `let` là `lazy-evaluated`: Tức là nó sẽ không được invoke khi hàm định nghĩa nó được invoke.
Ví dụ: 
```
$count = 0
describe "let" do
  let(:count) { $count += 1 }

  it "returns 1" do
    expect($count).to eq(1)
  end
end
```
Trong trường hợp này sẽ fail test trong rspec vì $count vẫn là 0 do hàm `count` chưa được invoke
Trường hợp sử dụng `let`: Trong trường hợp ta muốn 1 biến được cache trong 1 block hoặc khi ta muốn `lazy-evaluated` 
+ `let!` ngược lại vs let: Nó sẽ buộc hàm được thực thi 
Ví dụ 
```
$count = 0
describe "let!" do
  let!(:count) { $count += 1 }

  it "returns 1" do
    expect($count).to eq(1)
  end
end
```
Trong trường hợp này sẽ pass test trong rspec
Trường hợp sử dụng `let!`: Trong trường hợp ta muốn 1 biến được khởi tạo trước 1 khối `it` hoặc khi ta buộc hàm thực thi trước mỗi khối `it`
+ So sánh: Thông thường việc `lazy-evaluated` sẽ tiết kiệm được bộ nhớ hơn so với việc làm ngược lại. Nên việc sử dung `let` thường tiết kiệm bộ nhớ hơn